Add min available bytes storage policy

SamBarker opened the pull request:
First step in implementing [proposal #47](https://github.com/strimzi/proposals/blob/main/047-cluster-wide-volume-usage-quota-management.md).

This PR adds the first of the new limit types "min available bytes" and adds the requirement to specify the bootstrap URL for communicating with the Kafka cluster. It also removes the old storage checker, but preserves the old total consumed space soft and hard limits (as an incremental step we plan to remove them eventually) as well as their metrics.  

We propose to add additional metrics and other limit types in separate pull requests.

scholzj commented:
Looks like some spotbugs issues failed the build.

robobario commented on src/main/java/io/strimzi/kafka/quotas/StaticQuotaConfig.java:
```diff
@@ -4,37 +4,51 @@
+    static final String STORAGE_QUOTA_SOFT_PROP = CLIENT_QUOTA_CALLBACK_STATIC_PREFIX + ".storage.soft";
+    static final String STORAGE_QUOTA_HARD_PROP = CLIENT_QUOTA_CALLBACK_STATIC_PREFIX + ".storage.hard";
+    static final String STORAGE_CHECK_INTERVAL_PROP = CLIENT_QUOTA_CALLBACK_STATIC_PREFIX + ".storage.check-interval";
     static final String LOG_DIRS_PROP = "log.dirs";
```
This is dead code, we can remove it and it's getter etc

SamBarker replied:
I was thinking of keeping it to allow filtering the `logDirs` returned from the API, but your right it is currently dead code

SamBarker replied:
Resolved with c4ac2aab8e2399642327877534708381cff4de8b

robobario commented on src/main/java/io/strimzi/kafka/quotas/VolumeSourceBuilder.java:
```diff
@@ -0,0 +1,92 @@
+        this.adminClientFactory = adminClientFactory;
+    }
+
+     /*test*/ static Boolean testForKip827() {
```
We kept this in as a warning to anyone who tries to use this code with an old broker, even though we know for Strimzi it should only be used with compatible Kafka versions. Is this reasonable, or should we drop this code and put some warning in the README about compatibility?

SamBarker replied:
I was originally planning on making the migration story a bit cleaner, I think its worth keeping the check for now as it is a hard requirement for the plug-in

robobario commented on src/main/java/io/strimzi/kafka/quotas/AvailableBytesThrottleFactorSupplier.java:
```diff
@@ -0,0 +1,57 @@
+        boolean initial = throttled;
+        throttled = calculateNewFactor(volumes);
+        if (throttled != initial) {
+            listeners.forEach(Runnable::run);
```
we should debug log here when the throttle is opened/closed

SamBarker replied:
done in ee80af2d12467b4aedf8198a92e9439e3637e3c7

scholzj commented on pom.xml:
```diff
@@ -96,11 +96,12 @@
         <maven.enforced.version>3.0.0-M2</maven.enforced.version>
 
-        <kafka.version>3.0.0</kafka.version>
+        <kafka.version>3.3.1</kafka.version>
```
3.3.2? I guess we want to stick with the latest?

```suggestion
        <kafka.version>3.3.2</kafka.version>
```

scholzj commented on pom.xml:
```diff
@@ -117,6 +118,18 @@
         </repository>
     </distributionManagement>
 
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.junit</groupId>
+                <artifactId>junit-bom</artifactId>
+                <version>${junit.version}</version>
+                <scope>import</scope>
+                <type>pom</type>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
```
It seems weird to use `dependencyManagement` in a project with a single `pom.xml` and for a single dependency. What exactly is the reason?

SamBarker replied:
It allows us to import (note the use of `import`scope) the junit-bom and get consistent versions of all the junit dependencies. In this case both `junit-jupiter-api` and `junit-jupiter-params`. Checking its probably not a big deal for junit but comes in quite handy for things which have modules which move a bit more independently. 

Arguably overkill in this single pom but its just my default way (when available) for dealing with multi module imports from the same group.

scholzj replied:
Ok, good to know. Thanks for the explanation.

scholzj commented on pom.xml:
```diff
@@ -150,6 +168,26 @@
+        <dependency>
+            <groupId>com.github.spotbugs</groupId>
+            <artifactId>spotbugs-annotations</artifactId>
+            <version>4.7.2</version>
```
A nit, but having a property for the version as we do with others would be nice.

SamBarker replied:
have re-used the existing `spotbugs.version` property instead of introducing a new one as they were the same version.

scholzj reviewed (commented):
Left some (mostly useless) nit comments on the `pom.xml` ... I will probably need more time to go through the Java code yet, sorry 😞.

tombentley commented on src/main/java/io/strimzi/kafka/quotas/Volume.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.Objects;
+
+/**
+ * Represents a single volume on a specific Kafka broker.
```
I really represents the _usage_ of a volume on a broker. So maybe `VolumeUsage`?

SamBarker replied:
`VolumeUsage` is certainly better than the other names that we messed around with. I think I discounted `usage` as too ambiguous but thinking about it now there isn't really anything else it could realistically be referring to. 

Will update.

SamBarker replied:
Resolved with https://github.com/strimzi/kafka-quotas-plugin/pull/33/commits/824fd3e3a119b73d943f1375928da39564f88c86

tombentley commented on src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java:
```diff
@@ -145,49 +121,47 @@ public boolean updateClusterMetadata(Cluster cluster) {
-            backgroundScheduler.scheduleWithFixedDelay(storageChecker, storageCheckIntervalMillis, storageCheckIntervalMillis, TimeUnit.MILLISECONDS);
-            log.info("Configured quota callback with {}. Storage quota (soft, hard): ({}, {}). Storage check interval: {}ms", quotaMap, storageQuotaSoft, storageQuotaHard, storageCheckIntervalMillis);
+            Runnable volumeSource = volumeSourceBuilder.withConfig(config).withVolumeConsumer(throttleFactorSupplier).build();
+            backgroundScheduler.scheduleWithFixedDelay(volumeSource, 0, storageCheckIntervalMillis, TimeUnit.MILLISECONDS);
```
Although you've scheduled with fixed delay, the code that runs is actually only submitting requests to the `Admin` client. Those requests will be completed on the `Admin` client's completion thread. So if some of the requests you make using the Admin client take a long time to complete (e.g. requests start timing out) then it's possible the scheduler just keeps adding more and more requests to the `Admin`'s request queue.

Was that intentional? If so, it deserves Javadoc about why. If not, it might be simplest to just make `ClusterVolumeSource.run()` block by using `get` on the futures.

SamBarker replied:
Resolved with https://github.com/strimzi/kafka-quotas-plugin/pull/33/commits/139b97146eb3e6e61bdaefdce13998fa922a5986

tombentley reviewed (commented):
Thanks. I've made an first pass.

SamBarker commented on src/main/java/io/strimzi/kafka/quotas/VolumeSourceBuilder.java:
```diff
@@ -0,0 +1,92 @@
+ *
+ * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API">KIP-827 API</a>
+ */
+public class VolumeSourceBuilder implements AutoCloseable {
```
The disconnect in naming came from an attempt at backwards compatibility since abandoned originally the build returned a `LocalVolumeSource` or the `ClusterVolumeSource` depending on the configuraiton. Will clean up the naming now.

SamBarker replied:
resolved with https://github.com/strimzi/kafka-quotas-plugin/pull/33/commits/7d4630fe4b90ebe8638a4a79d1109f7755ad0589

SamBarker commented on src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java:
```diff
@@ -145,49 +121,47 @@ public boolean updateClusterMetadata(Cluster cluster) {
+            backgroundScheduler.scheduleWithFixedDelay(volumeSource, 0, storageCheckIntervalMillis, TimeUnit.MILLISECONDS);
+            log.info("Configured quota callback with {}. Storage check interval: {}ms", quotaMap, storageCheckIntervalMillis);
+        } else {
+            log.warn("Static quota callback configured to never check usage: set {} to a positive value to enable", StaticQuotaConfig.STORAGE_CHECK_INTERVAL_PROP);
```
🤷 The plug-in has been explicitly added to the class path which suggests an intent to monitor but wont do anything as its missing configuration enabling it. 

I'd be happy with either of `WARN` or `INFO`, but landed on `WARN` as that is what the text of the message really is and as it stands out in the logs more.

tombentley replied:
It sounds like this is really, in effect, a required config parameter (if the callback is being used, which be definition is is because `configure()` is being called). If so, then either it should be an `error`, or you should just throw an exception (probably `ConfigException`) with a suitable message.

SamBarker replied:
This change is just to add the logging when it's turned off. The actual check and ability to [disable the plug-in](https://github.com/strimzi/kafka-quotas-plugin/blob/main/src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java#L163-L170) is existing behaviour. 

So we can remove support for disabling the plug-in, by enforcing a positive non 0 [value here](https://github.com/strimzi/kafka-quotas-plugin/blob/fc5babec5f77b2090b7633d4bb0d819c8bea7147/src/main/java/io/strimzi/kafka/quotas/StaticQuotaConfig.java#L63l). You can still effectively disable the plug-in by setting it to `Int.MAX_VALUE` though that would still start thread-pools etc if we care about the distinction.

SamBarker replied:
The other point I've just realised, and mentioned elsewhere but worth re-iterating here, the plug-in could be used without monitoring disk usage to just apply static quotas. So keeping support for disabling disabling the usage monitoring seems appropriate.

Though it does also mean we should probably drop the log level to `INFO` rather than `WARN` as its a valid configuration.

SamBarker replied:
resolved with 86540161ae6062f009eb32af4455b707df179ec7

SamBarker commented on src/main/java/io/strimzi/kafka/quotas/ThrottleFactorSupplier.java:
```diff
@@ -0,0 +1,25 @@
+ * Where a value of `1.0` implies no additional restriction over and above the defined quota.
+ * A value of `0.0` implies that there is no quota available regardless of the defined quota.
+ */
+public interface ThrottleFactorSupplier extends Supplier<Double>, Consumer<Collection<Volume>> {
```
`ThrottleFactorPolicy` is definitely a better name. 

Will experiment with the other options to see what feels better.

SamBarker replied:
Rename done with https://github.com/strimzi/kafka-quotas-plugin/pull/33/commits/20577877d8ce53fde63acfbe104a035f2ea28fbd

SamBarker replied:
I've got with 2.2 in 320e10b92c977c719d23e47d0cf953b9f3ead8b4 and removed the use of Supplier in ef96f1915e9c1a2c7ed72d11455f4a22dd2838da. 

I've stuck with a dedicated interface for now as I think there might be value later in having a separate observer for metrics, which is something I plan to handle in a separate PR, if they don't end up making sense as a separate consumer I'll probably collapse the interface then.

SamBarker commented on src/main/java/io/strimzi/kafka/quotas/VolumeSourceBuilder.java:
```diff
@@ -0,0 +1,92 @@
+     * @param kip827Available used to determine if KIP-827 API's are available
+     * @param adminClientFactory factory function for creating Admin clients with the builders' config.
+     */
+    /* test */ VolumeSourceBuilder(Supplier<Boolean> kip827Available, Function<StaticQuotaConfig.KafkaClientConfig, Admin> adminClientFactory) {
```
I was delaying the evaluation until build time as that would only be executed if the plug-in was actually configured to run and thus the API availability actually matters. 

The delayed execution point is worth documenting, assuming we keep it.

tombentley replied:
Why do you need delayed evaluation though? You're _at most_ loading a class (`LogDirDescription`) and reflecting a single method. That's static information, so the answer is not going to change. And the cost is not _so huge_ that it's worth deferring until you know you need the result.

SamBarker replied:
We need delayed execution if we continue to support having the plug-in loaded but disabled as I'd argue if it's disabled it shouldn't explode on with the missing api.

Maybe that's better handled by restructuring the construction logic rather than delaying the test.

SamBarker replied:
So I've had a look at restructuring it so the test can be done once and avoid the supplier. The trouble is we construct/inject the `VolumeSourceBuilder` during construction of the `StaticQuotaCallback` which is very helpful for unit testing. 

However that does mean we can't invoke the test during construction as that would prevent the plug-in loading, I can't remember what that does to broker startup, which would be annoying if users have it disabled. I don't know how safe it is to assume that the plug-in is running in a 3.3+ broker (especially as there is no reason it has to run only in Strimzi deployments).

I'm of the opinion its better to delay the execution of the test until its required even if the answer doesn't change (we could use [Guava memoize](https://guava.dev/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier), iirc guava is on the class path, to make that more explicit but doesn't seem worth it to me) rather than prevent everything loading for older brokers. 

The plug-in could be used without monitoring disk usage to just apply static quotas, so failing construction is probably not the appropriate choice.

SamBarker replied:
@robobario and I have gone for a wider ranging refactor than I looked at before and have removed the supplier while preserving the delayed evaluation with 7433cc641160300490017cd06fb3f8a301c57a3c

SamBarker commented on src/main/java/io/strimzi/kafka/quotas/VolumeSourceBuilder.java:
```diff
@@ -0,0 +1,92 @@
+        return this;
+    }
+
+    Runnable build() {
```
Resolved in https://github.com/strimzi/kafka-quotas-plugin/pull/33/commits/7d4630fe4b90ebe8638a4a79d1109f7755ad0589

robobario commented on src/main/java/io/strimzi/kafka/quotas/ThrottleFactorSupplier.java:
```diff
@@ -0,0 +1,25 @@
+     * Register a listener to be invoked when the throttle factor is changed.
+     * @param listener to be executed whenever the factor is updated.
+     */
+    void addUpdateListener(Runnable listener);
```
We also need to support the static quotas for non-PRODUCE quota types, so we could have an awkward split where the non-PRODUCE quota types are handled by the existing set and PRODUCE quota types have to delegate to the throttle supplier. It might be simpler as-is to have it tell the `StaticQuotaCallback` that PRODUCE has changed so they can share the same mechanism. Though as you said the details could be cleaned up.

I also like how currently `StaticQuotaCallback` contains the knowledge about this `quotaResetRequired` API (along with it's performance requirements). Having an atomic boolean in the `ThrottleFactorSupplier` somewhere that is polled by  StaticQuotaCallback using `getAndSet(false)` implies that the `ThrottleFactorSupplier` would be aware of this quota-reset API. Maybe that's okay, if we think it makes things more readable the `ThrottleFactorSupplier` could implement `public boolean quotaResetRequired(ClientQuotaType quotaType)` but it becomes less about naively handing off the factor and more coupled to the callback API.

Another note is we went with the `setListener` style since we've ended up with two ThrottleFactorPolicy implementations and wanted a common way to wire in the listener (since telling StaticQuotaCallback was a requirement for both). A setter was something we could put in the interface. Maybe we can refactor it further so there is a single class responsible for knowing the factor has changed and telling the StaticQuotaCallback. I imagine we can get it to the point where we use the constructor to pass a single listener instead of using this setter. We could introduce an interface instead of Runnable to document that it shouldn't block up the world since it's called from the admin thread.

SamBarker replied:
@robobario and I think 0e91ea08b8ae84a0ab1fcfa218097d69eb4eed3d addresses the spirit of this conversation.

tombentley commented on src/main/java/io/strimzi/kafka/quotas/AvailableBytesThrottleFactorPolicy.java:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private boolean calculateNewFactor(Collection<VolumeUsage> volumes) {
```
This doesn't actually calculate a factor. Maybe call it `shouldThrottle`?

SamBarker replied:
good point resolved with calculateNewFactor no longer calculates so give it a better name :D Sam Barker A minute ago

tombentley commented on src/main/java/io/strimzi/kafka/quotas/ThrottleFactorPolicy.java:
```diff
@@ -0,0 +1,26 @@
+    /**
+     * @return The current factor in the range {@code [0..1]} to scale the throttle by.
+     */
+    Double currentFactor();
```
None of the existing policies seem to require null, and the meaning of null is not defined, so why is this not a primitive `double`?

SamBarker replied:
resolved with 434502c53e82bd61cbc8ad3a82a7cd9a0dbaf431

tombentley commented on src/main/java/io/strimzi/kafka/quotas/VolumeSource.java:
```diff
@@ -0,0 +1,119 @@
+        final DescribeClusterResult clusterResult = admin.describeCluster();
+        try {
+            clusterResult.nodes().whenComplete((nodes, throwable) -> {
+                if (throwable != null) {
+                    log.error("error while describing cluster", throwable);
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Successfully described cluster: " + nodes);
+                    }
+                    onClusterDescribeSuccess(nodes);
+                }
```
This is still being executed on the Admin client's response handler thread. Ultimately the call to `volumeObserver.observeVolumeUsage(volumes);` also runs on the Admin client's response handler thread, so it ought not to block, but we don't actually know how long it might block for (because some of the factor policies invoke `Runnable#run` in `observeVolumeUsage`). Now, I know _in practice_ we're not expecting it to block, because the `listener.run()` is just adding something to a collection. 

TBH I think it's just better to avoid using `whenComplete()` _at all_ in this class, and just write get the Admin results by blocking on the thread using `get`. I think that would give you what you actually intend by the scheduled execution of this `run()`. It also simpler code.

SamBarker replied:
Resolved with 876b9a0b102c96ea8c946b3d471684472c109976

tombentley reviewed (commented):
Thanks, I think the intent and the code is a lot clearer now. I left a few more comments.

SamBarker commented:
@tombentley: @robobario and I think we have addressed all your feedback hopefully the changes work for you. 

(cc: @scholzj if you are still planning to look at the java code)

tombentley commented on src/main/java/io/strimzi/kafka/quotas/PolicyBasedThrottle.java:
```diff
@@ -0,0 +1,45 @@
+
+    private final ThrottleFactorPolicy factorPolicy;
+    private final Runnable listener;
+    private volatile Double throttleFactor = 1.0d;
```
This can be a primitive

SamBarker replied:
resolved with ed052f6ffe7180ae07791c5dcf4dff76dd8007ab

tombentley commented on src/main/java/io/strimzi/kafka/quotas/PolicyBasedThrottle.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Override
+    public void observeVolumeUsage(Collection<VolumeUsage> observedVolumes) {
+        Double old = this.throttleFactor;
```
This can be a primitive

SamBarker replied:
resolved with ed052f6ffe7180ae07791c5dcf4dff76dd8007ab

tombentley reviewed (approved):
A few more nits, but assuming you're happy to fix those this LGTM. Thanks!

SamBarker commented on src/test/java/io/strimzi/kafka/quotas/AvailableBytesThrottleFactorPolicyTest.java:
```diff
@@ -0,0 +1,79 @@
+        final double actualFactor = availableBytesThrottleFactorSupplier.calculateFactor(List.of(volumeWithAvailableBytes(1000L)));
+
+        //Then
+        assertThat(actualFactor).isCloseTo(1.0d, Offset.offset(0.00001));
```
resolved with 79a180c83ef86ebc00bd99d09890805b933599ef

SamBarker commented on src/main/java/io/strimzi/kafka/quotas/TotalConsumedThrottleFactorPolicy.java:
```diff
@@ -0,0 +1,76 @@
+        if (totalConsumed >= consumedBytesHardLimit) {
+            return  0.0;
+        } else if (totalConsumed >= consumedBytesSoftLimit) {
+            return 1.0d - (1.0d * (totalConsumed - consumedBytesSoftLimit) / (consumedBytesHardLimit - consumedBytesSoftLimit));
```
resolved with 103181e452203ef8f54badafe1f70122e44a375a.

Ended going a slightly different direction and pulling out field for the denominator and a variable for the numerator as that added more descriptive names as well as type information, its also a tiny optimisation as the range can be calculated on construction (in practice I think the JIT would spot that anyway).

SamBarker commented on src/main/java/io/strimzi/kafka/quotas/VolumeUsage.java:
```diff
@@ -0,0 +1,94 @@
+     * @param capacity How many bytes the volume holds
+     * @param availableBytes How many available bytes remain on the volume.
+     */
+    public VolumeUsage(String brokerId, String logDir, long capacity, long availableBytes) {
```
resolved with fa5bfa3e846272faeeba42aab294147e059157b4

SamBarker commented on src/main/java/io/strimzi/kafka/quotas/AvailableBytesThrottleFactorPolicy.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private boolean shouldThrottle(Collection<VolumeUsage> observedVolumes) {
+        return observedVolumes.stream().anyMatch(volume -> volume.getAvailableBytes() <= availableBytesLimit);
```
We plan to address data validity concerns as part of a separate PR to add the [fallback factor](https://github.com/strimzi/proposals/blob/main/047-cluster-wide-volume-usage-quota-management.md#throttle-factor-fallback). So for now we have added filtering to the VolumeSource to prevent it passing on volumes without reported usage.

resolved with fa5bfa3e846272faeeba42aab294147e059157b4

SamBarker commented on src/test/java/io/strimzi/kafka/quotas/AvailableBytesThrottleFactorPolicyTest.java:
```diff
@@ -0,0 +1,79 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+@ExtendWith(MockitoExtension.class)
+class AvailableBytesThrottleFactorPolicyTest {
```
Have added a test in fa5bfa3e846272faeeba42aab294147e059157b4 which demonstrates the filtering behaviour.

scholzj reviewed (approved):
LGTM I guess ... assuming it works :-D

scholzj commented:
@ppatierno @strimzi/maintainers Anyone has anything else to this?
